File: android/manager/src/main/java/com/dreampany/manager/ui/fragment/HomeFragment.java
package com.dreampany.manager.ui.fragment;

import android.os.Bundle;

import com.dreampany.manager.R;
import com.dreampany.manager.ui.model.UiTask;
import com.dreampany.framework.data.model.Task;
import com.dreampany.framework.misc.ActivityScope;
import com.dreampany.framework.ui.fragment.BaseFragment;
import com.dreampany.framework.ui.fragment.BaseStateFragment;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;


/**
 * Created by Hawladar Roman on 6/20/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
@ActivityScope
public class HomeFragment extends BaseStateFragment<BaseFragment> {

    @Inject
    ViewModelProvider.Factory factory;

    @Inject
    public HomeFragment() {
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_tabpager_fixed;
    }

    @Override
    public int getMenuId() {
        return R.menu.menu_home;
    }

    @Override
    public int getSearchMenuItemId() {
        return R.id.item_search;
    }

    @NonNull
    @Override
    protected String[] pageTitles() {
        return new String[]{"","", ""};
    }

    @NonNull
    @Override
    protected Class<BaseFragment>[] pageClasses() {
        return new Class[]{FirstFragment.class, SecondFragment.class, ThirdFragment.class};
    }

    @NotNull
    @Override
    protected Task<?>[] pageTasks() {
        UiTask<?> task = getCurrentTask(false);
        return new Task[]{task, task, task};
    }

    @Override
    public boolean hasAllPages() {
        return true;
    }

    @Override
    public boolean hasTabColor() {
        return true;
    }

    @Override
    protected void onStartUi(@Nullable Bundle state) {
        initView();
    }

    @Override
    protected void onStopUi() {

    }


    @NonNull
    @Override
    public ViewModelProvider.Factory get() {
        return factory;
    }

    @Override
    public boolean onQueryTextChange(@NonNull String newText) {
        BaseFragment fragment = getCurrentFragment();
        return fragment != null && fragment.onQueryTextChange(newText);
    }

    private void initView() {
        setTitle(R.string.home);
    }
}

File: android/cast/src/main/kotlin/com/dreampany/cast/ui/fragment/HomeFragment.kt
package com.dreampany.cast.ui.fragment

import android.Manifest
import android.os.Bundle
import androidx.databinding.ObservableArrayList
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.dreampany.cast.R
import com.dreampany.cast.databinding.ContentRecyclerBinding
import com.dreampany.cast.databinding.FragmentHomeBinding
import com.dreampany.cast.ui.adapter.UserAdapter
import com.dreampany.cast.vm.UserViewModel
import com.dreampany.frame.data.enums.UiState
import com.dreampany.frame.databinding.ContentTopStatusBinding
import com.dreampany.frame.misc.ActivityScope
import com.dreampany.frame.ui.fragment.BaseMenuFragment
import com.dreampany.frame.util.ViewUtil
import com.karumi.dexter.MultiplePermissionsReport
import eu.davidea.flexibleadapter.common.FlexibleItemDecoration
import eu.davidea.flexibleadapter.common.SmoothScrollLinearLayoutManager
import javax.inject.Inject

/**
 * Created by Roman-372 on 7/16/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@ActivityScope
class HomeFragment @Inject constructor() : BaseMenuFragment() {

    @Inject
    lateinit var factory: ViewModelProvider.Factory
    private lateinit var bindHome: FragmentHomeBinding
    private lateinit var bindStatus: ContentTopStatusBinding
    private lateinit var bindRecycler: ContentRecyclerBinding

    private lateinit var refresh: SwipeRefreshLayout

    private lateinit var adapter: UserAdapter
    private lateinit var vm: UserViewModel

    override fun getLayoutId(): Int {
        return R.layout.fragment_home
    }

    override fun onStartUi(state: Bundle?) {
        initView()
        initRecycler()
        checkPermissions(Manifest.permission.ACCESS_FINE_LOCATION
        )
    }

    override fun onStopUi() {
        vm.stopNearby()
    }

    override fun onPermissionsChecked(report: MultiplePermissionsReport) {
        if (report.areAllPermissionsGranted()) {
            vm.startNearby()
        }
    }

    private fun initView() {
        setTitle(R.string.home)
        bindHome = super.binding as FragmentHomeBinding
        bindStatus = bindHome.layoutTopStatus
        bindRecycler = bindHome.layoutRecycler

        refresh = bindHome.layoutRefresh

        vm = ViewModelProviders.of(this, factory).get(UserViewModel::class.java)
        vm.observeUiState(this, Observer { processUiState(it) })
    }

    private fun initRecycler() {
        bindHome.items = ObservableArrayList<Any>()
        adapter = UserAdapter(this)
        adapter.setStickyHeaders(false)
        ViewUtil.setRecycler(
            adapter,
            bindRecycler.recycler,
            SmoothScrollLinearLayoutManager(context!!),
            FlexibleItemDecoration(context!!)
                .addItemViewType(R.layout.item_user, vm.itemOffset)
                .withEdge(true),
            null,
            null,
            null
        )
    }

    private fun processUiState(state: UiState) {
        when (state) {
            UiState.SHOW_PROGRESS ->
                if (!refresh.isRefreshing) {
                    refresh.isRefreshing = true
                }
            UiState.HIDE_PROGRESS ->
                if (refresh.isRefreshing) {
                    refresh.isRefreshing = false
                }
            UiState.OFFLINE -> {
            }
            UiState.ONLINE -> {
            }
            UiState.EXTRA -> processUiState(if (adapter.isEmpty) UiState.EMPTY else UiState.CONTENT)
            UiState.SEARCH -> {
            }
            UiState.EMPTY -> {
            }
            UiState.ERROR -> {
            }
            UiState.CONTENT -> {
            }
        }// bindStatus.layoutExpandable.expand();
        //   bindStatus.layoutExpandable.collapse();
        //binding.stateful.setState(SEARCH);
        //   binding.stateful.setState(SEARCH);
        //  binding.stateful.setState(StatefulLayout.State.CONTENT);
    }
}